### PR TITLE
FeatureFinderMultiplex update

### PR DESCRIPTION
**Pull request by lars20070**
The old version of `FeatureFinderMultiplex` filtered `mzML` spectrum by spectrum. Single spectra are now replaced by *spectra bundles*, see new parameter `rt_band`. This has the advantage that peaks in neighbouring spectra can be used to complete an isotopic pattern, i.e. the `SpectraMerger` tool is no longer needed. This has several advantages:

(1) Higher sensitivity at low intensities.
(2) Complete mass traces
(3) Detection of RT shifts in peptide pairs possible, e.g. in SILAC heavy elutes before light.

For `rt_band = 0` the feature finder behaves like in the old version.

screen shot 2018-04-04 at 13 45 28

Here an example of the increased sensitivity and missing peaks.

screen shot 2018-04-04 at 13 55 21

### CODE REVIEW

**Comment by lars20070** [comment]
Test updates coming soon.

**Comment by hroest** [comment]
this definitely should have some black-box tests. Also the algorithm in the tool could be moved into OpenMS as a FFMultiplex Algorithm. Finally I suggest to also update the pyOpenMS tests, see https://github.com/OpenMS/OpenMS/wiki/pyOpenMS#testing

**Comment by hroest** [comment]
can you please fix:

- http://cdash.openms.de/viewTest.php?onlyfailed&buildid=264775
- http://cdash.openms.de/viewBuildError.php?buildid=264774
- http://cdash.openms.de/viewBuildError.php?buildid=264773

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.h [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+   * @brief generate list of m/z shifts
+   *
+   * @param charge_min    minimum charge
+   * @param charge_max    maximum charge
+   * @param peaks_per_peptide_max    maximum number of isotopes in peptide
+   * @param mass_pattern_list    mass shifts due to labelling
+   *
+   * @return list of m/z shifts
+   */
+  std::vector<MultiplexIsotopicPeakPattern> generatePeakPatterns_(int charge_min, int charge_max, int peaks_per_peptide_max, std::vector<MultiplexDeltaMasses> mass_pattern_list);
```
Comment (hroest) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.h]: can you pass the vector by (const) reference?

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.h [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  
+  /**
+   * @brief calculate peptide intensities
+   *
+   * @param pattern
+   * @param satellites
+   *
+   * @return vector with intensities for each of the peptides
+   */
+  std::vector<double> determinePeptideIntensitiesCentroided_(MultiplexIsotopicPeakPattern& pattern, std::multimap<size_t, MultiplexSatelliteCentroided >& satellites);
```
Comment (hroest) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.h]: use const-ref?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredMSExperiment.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      consensus.setMZ(it_peak->getMZ());
+      consensus.setIntensity(1.0);
+      consensus.setCharge(1);
+      consensus.setQuality(1.0);
+      
+      std::multimap<size_t, MultiplexSatelliteCentroided > satellites = it_peak->getSatellites();
+      int count = 0;
+      
+      // loop over satellites
+      for (std::multimap<size_t, MultiplexSatelliteCentroided >::const_iterator it_satellite = satellites.begin(); it_satellite != satellites.end(); ++it_satellite)
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredMSExperiment.cpp]: you could make this a bit more readable: `for (const auto & it_satellite : satellites) `

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredPeak.cpp [hunk 4]
```diff
@@ -32,88 +32,82 @@
...
-     */
-    std::vector<MultiplexFilterResultRaw> raw_data_points_;
- 
-  };
+  void MultiplexFilteredPeak::addSatellite(size_t rt_idx, size_t mz_idx, size_t pattern_idx)
+  {
+    satellites_.insert(std::make_pair(pattern_idx, MultiplexSatelliteCentroided(rt_idx, mz_idx)));
+  }
+  
+  void MultiplexFilteredPeak::addSatellite(MultiplexSatelliteCentroided satellite, size_t pattern_idx)
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredPeak.cpp]: `void MultiplexFilteredPeak::addSatellite(const MultiplexSatelliteCentroided& satellite, size_t pattern_idx) `

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredPeak.cpp [hunk 5]
```diff
@@ -32,88 +32,82 @@
...
+  {
+    satellites_.insert(std::make_pair(pattern_idx, satellite));
+  }
+  
+  void MultiplexFilteredPeak::addSatelliteProfile(double rt, double mz, double intensity, size_t pattern_idx)
+  {
+    satellites_profile_.insert(std::make_pair(pattern_idx, MultiplexSatelliteProfile(rt, mz, intensity)));
+  }
   
+  void MultiplexFilteredPeak::addSatelliteProfile(MultiplexSatelliteProfile satellite, size_t pattern_idx)
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredPeak.cpp]: const-ref

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 6]
```diff
@@ -34,388 +34,590 @@
...
-  {
-  }
-
-  int MultiplexFiltering::positionsAndBlacklistFilter_(const MultiplexIsotopicPeakPattern& pattern, int spectrum,
-                                                      const vector<double>& peak_position, int peak,
-                                                      vector<double>& mz_shifts_actual,
-                                                      vector<int>& mz_shifts_actual_indices) const
+  MultiplexFiltering::MultiplexFiltering(const MSExperiment& exp_picked, const std::vector<MultiplexIsotopicPeakPattern> patterns, int isotopes_per_peptide_min, int isotopes_per_peptide_max, double intensity_cutoff, double rt_band, double mz_tolerance, bool mz_tolerance_unit, double peptide_similarity, double averagine_similarity, double averagine_similarity_scaling, String averigine_type) :
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: `const std::vector<MultiplexIsotopicPeakPattern>& patterns`

also, please keep line lengths < 100 chars

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 7]
```diff
@@ -34,388 +34,590 @@
...
-    }
+      blacklist_.push_back(blacklist_spectrum);
+    }    
+  }
 
-    // count how many isotopic peaks seen simultaneously in all of the peptides
-    // and (optionally) remove peaks following missing ones
-    int peaks_found_in_all_peptides = peaks_per_peptide_max_;
-    for (int peptide = 0; peptide < (int) pattern.getMassShiftCount(); ++peptide)
+  MSExperiment MultiplexFiltering::getWhiteMSExperiment_(White2Original& mapping)
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: use const-ref

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 8]
```diff
@@ -34,388 +34,590 @@
...
-
-  bool MultiplexFiltering::averagineSimilarityFilter_(const MultiplexIsotopicPeakPattern& pattern, const vector<double>& intensities_actual, int peaks_found_in_all_peptides_spline, double mz) const
-  {
-    // Use a more restrictive averagine similarity when we are searching for peptide singlets.
-    double similarity;
-    if (pattern.getMassShiftCount() == 1)
+    
+    // determine absolute m/z tolerance in Th
+    double mz_tolerance;
+    if (mz_tolerance_unit_)
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: I suggest to rename this to `mz_tolerance_unit_in_ppm_`

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringCentroided.cpp [hunk 9]
```diff
@@ -38,246 +38,144 @@
...
-
 using namespace std;
 using namespace boost::math;
 
 namespace OpenMS
 {
 
-  MultiplexFilteringCentroided::MultiplexFilteringCentroided(const PeakMap& exp_picked, const std::vector<MultiplexIsotopicPeakPattern> patterns, int peaks_per_peptide_min, int peaks_per_peptide_max, bool missing_peaks, double intensity_cutoff, double mz_tolerance, bool mz_tolerance_unit, double peptide_similarity, double averagine_similarity, double averagine_similarity_scaling, String averagine_type) :
-    MultiplexFiltering(exp_picked, patterns, peaks_per_peptide_min, peaks_per_peptide_max, missing_peaks, intensity_cutoff, mz_tolerance, mz_tolerance_unit, peptide_similarity, averagine_similarity, averagine_similarity_scaling, averagine_type)
+  MultiplexFilteringCentroided::MultiplexFilteringCentroided(const MSExperiment& exp_picked, const std::vector<MultiplexIsotopicPeakPattern> patterns, int isotopes_per_peptide_min, int isotopes_per_peptide_max, double intensity_cutoff, double rt_band, double mz_tolerance, bool mz_tolerance_unit, double peptide_similarity, double averagine_similarity, double averagine_similarity_scaling, String averagine_type) :
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringCentroided.cpp]: pass patterns by const-ref

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp [hunk 10]
```diff
@@ -34,105 +34,93 @@
...
-    MultiplexFiltering(exp_picked, patterns, peaks_per_peptide_min, peaks_per_peptide_max, missing_peaks, intensity_cutoff, mz_tolerance, mz_tolerance_unit, peptide_similarity, averagine_similarity, averagine_similarity_scaling, averagine_type), exp_profile_(exp_profile), boundaries_(boundaries)
+  MultiplexFilteringProfile::MultiplexFilteringProfile(MSExperiment& exp_profile, const MSExperiment& exp_picked, const std::vector<std::vector<PeakPickerHiRes::PeakBoundary> >& boundaries, const std::vector<MultiplexIsotopicPeakPattern> patterns, int isotopes_per_peptide_min, int isotopes_per_peptide_max, double intensity_cutoff, double rt_band, double mz_tolerance, bool mz_tolerance_unit, double peptide_similarity, double averagine_similarity, double averagine_similarity_scaling, String averagine_type) :
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: pass patterns by const-ref

Comment (hroest) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: can you make `exp_profile` const?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp [hunk 11]
```diff
@@ -141,231 +129,308 @@ namespace OpenMS
...
+    // clock for monitoring run performance
+    //std::cout << "\nThat took me " << (float)(clock()-start)/CLOCKS_PER_SEC << " seconds.\n";
 
     endProgress();
 
     return filter_results;
   }
 
-  int MultiplexFilteringProfile::nonLocalIntensityFilter_(const MultiplexIsotopicPeakPattern& pattern, const vector<double>& mz_shifts_actual, const vector<int>& mz_shifts_actual_indices, SplineSpectrum::Navigator nav, std::vector<double>& intensities_actual, int peaks_found_in_all_peptides, double mz) const
+  bool MultiplexFilteringProfile::filterAveragineModel_(const MultiplexIsotopicPeakPattern& pattern, const MultiplexFilteredPeak& peak, std::multimap<size_t, MultiplexSatelliteProfile > satellites_profile) const
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: make `satellites_profile` const-ref?

**Comment by hroest** [comment]
also, can you add unit tests for the new classes?

**Comment by lars20070** [comment]
Currently progress is not reported (set by the `no_progress` flag). In the olf `FFM` tool it worked alright.

Is there an algorithm class where progress works right now and from which I can copy?

**Comment by cbielow** [comment]
Probably because logging is not forwarded to the commandline?
You need to use `youralgoObj.setLogType(this->log_type_);` in the TOPP tool, e.g.
https://github.com/OpenMS/OpenMS/blob/develop/src/topp/PeptideIndexer.cpp#L143

**Comment by lars20070** [comment]
Yep, [that's](https://github.com/OpenMS/OpenMS/blob/develop/src/topp/PeptideIndexer.cpp#L143) it. Thanks @cbielow.

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredMSExperiment.cpp (hunk 3 above)
Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredMSExperiment.cpp]: @hroest Thanks for pointing this out. Very neat.

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp [hunk 12]
```diff
@@ -0,0 +1,1114 @@
...
+     std::cout << "charge = " << list[i].getCharge() << "+    shift = " << list[i].getMassShiftAt(1) << " Da\n";
+     }*/
+    
+    return list;
+  }
+  
+  /**
+   * @brief simple linear regression through the origin
+   *
+   * TODO: combine with OpenMS/MATH/STATISTICS/LinearRegression
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp]: I would suggest a separate class `OpenMS/MATH/STATISTICS/LinearRegressionWithoutIntercept` in order to keep the interface simple.

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp (hunk 7 above)
Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: `mapping` is being returned, so cannot be `const`.

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 13]
```diff
@@ -34,388 +34,590 @@
...
         return true;
       }
     }
 
     return false;
   }
-
-  bool MultiplexFiltering::peptideSimilarityFilter_(const MultiplexIsotopicPeakPattern& pattern, const vector<double>& intensities_actual, int peaks_found_in_all_peptides_spline) const
+  
+  bool MultiplexFiltering::filterPeakPositions_(const MSSpectrum::ConstIterator& it_mz, White2Original& index_mapping, const MSExperiment::ConstIterator& it_rt_begin, const MSExperiment::ConstIterator& it_rt_band_begin, const MSExperiment::ConstIterator& it_rt_band_end, const MultiplexIsotopicPeakPattern& pattern, MultiplexFilteredPeak& peak) const
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: Will require new `const []` operator for `White2Original`.
Can we do without?

**Comment by timosachsenberg** [comment]
Use .at() instead of const?

**Comment by lars20070** [comment]
No. Complains at the same `at()`.
```
/Users/lars/Code/OpenMS-develop/OpenMS/src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp:222:89: error: no viable overloaded operator[] for type
      'const std::__1::__vector_base<std::__1::map<int, int, std::__1::less<int>, std::__1::allocator<std::__1::pair<const int, int> > >,
      std::__1::allocator<std::__1::map<int, int, std::__1::less<int>, std::__1::allocator<std::__1::pair<const int, int> > > > >::value_type' (aka 'const
      std::__1::map<int, int, std::__1::less<int>, std::__1::allocator<std::__1::pair<const int, int> > >')
            peak.addSatellite(it_rt - it_rt_begin, index_mapping.at(it_rt - it_rt_begin)[i], idx_mz_shift);
```

**Comment by lars20070** [comment]
@timosachsenberg Yep, works. Thanks.

File: src/openms/include/OpenMS/FILTERING/DATAREDUCTION/SplineSpectrum.h [hunk 14]
```diff
@@ -93,7 +93,7 @@ class OPENMS_DLLAPI SplineSpectrum
      *  has some usable data to work on.
      *  In case there are no packages, a subsequent call to getNavigator() will throw an exception.
      */
-    size_t getSplineCount() const;
+    size_t size() const;
```
Comment (hroest) [inline src/openms/include/OpenMS/FILTERING/DATAREDUCTION/SplineSpectrum.h]: you will need to change the corresping `.pxd` file

**Comment by hroest** [comment]
@lars20070 can you also fix the pyOpenMS errors: http://cdash.openms.de/viewBuildError.php?buildid=267705

**Comment by lars20070** [comment]
@hroest Will do.

Currently working on the TOPP tests. [This](https://github.com/lars20070/OpenMS/blob/b1a93ddba19dedba0690c26c43df899ef451894f/src/tests/topp/CMakeLists.txt#L279) line of the test fails (the execution, not the comparison of the outputs). But both on macOS and Ubuntu this command runs through fine.

File: src/openms/include/OpenMS/MATH/STATISTICS/StatisticFunctions.h [hunk 15]
```diff
@@ -72,108 +70,110 @@ namespace OpenMS
...
-      std::sort(begin, end);
+      return std::accumulate(begin, end, 0.0);
     }
 
-    Size size = std::distance(begin, end);
-    if (size % 2 == 0)        // even size => average two middle values
+    /**
+       @brief Calculates the mean of a range of values
+      
+       @exception Exception::InvalidRange is thrown if the range is NULL
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/MATH/STATISTICS/StatisticFunctions.h]: A range is not NULL. A range is empty

File: src/openms/include/OpenMS/MATH/STATISTICS/StatisticFunctions.h [hunk 16]
```diff
@@ -72,108 +70,110 @@ namespace OpenMS
...
     }
-    else
+
+    /**
+       @brief Calculates the median of a range of values
+       
+       @param begin Start of range
+       @param end End of range (past-the-end iterator)
+       @param sorted Is the range already sorted? If not, it will be sorted.
+    @return Median (as floating point, since we need to support average of middle values)
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/MATH/STATISTICS/StatisticFunctions.h]: indentation?

File: src/openms/include/OpenMS/MATH/STATISTICS/StatisticFunctions.h (hunk 15 above)
Comment (lars20070) [reply src/openms/include/OpenMS/MATH/STATISTICS/StatisticFunctions.h]: `StatisticFunctions` changes are not from me. That's why I removed it again from this PR.

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredMSExperiment.h [hunk 17]
```diff
@@ -46,37 +45,37 @@ namespace OpenMS
...
          * @brief constructor
          */
-        MultiplexFilterResult();
+        MultiplexFilteredMSExperiment();
         
         /**
          * @brief adds a single peak to the results
          */
-        void addFilterResultPeak(double mz, double rt, std::vector<double> mzShifts, std::vector<double> intensities, std::vector<MultiplexFilterResultRaw> result);
+        void addPeak(double mz, double rt, size_t mz_idx, size_t rt_idx);
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredMSExperiment.h]: It's an alternative/shorter way to add a filtered peak. Currently not used, but useful in the API.

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredPeak.h [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+    
+    /**
+     * @brief return all satellite data points
+     */
+    const std::multimap<size_t, MultiplexSatelliteProfile >& getSatellitesProfile() const;
+    
+    /**
+     * @brief return number of satellite peaks
+     */
+    size_t size() const;
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteredPeak.h]: https://github.com/OpenMS/OpenMS/pull/3353/files#diff-fe48f655988461c32c87888af4e4850aR62

In both modes, centroided data are stored in these results (`size()`). In profile mode, the spline-interpolated profile data points are stored additionally (`sizeProfile()`).

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.h [hunk 19]
```diff
@@ -73,187 +74,155 @@ namespace OpenMS
...
+     * spectrum back to its original spectrum.
      */
-    struct BlackListEntry
-    {
-      bool black;
-      int black_exception_mass_shift_index;
-      int black_exception_charge;
-      int black_exception_mz_position;
-    };
+    typedef std::vector<std::map<int, int> > White2Original;
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.h]: No. I filter small, white (i.e. non-blacklisted) experiments but then blacklist in the original input experiment. I need to quickly move between experiments and do this with the [map](https://github.com/OpenMS/OpenMS/pull/3353/files#diff-d07005aa96a93e57990ba6d481c8425eR224).

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexSatelliteProfile.cpp [hunk 20]
```diff
@@ -32,31 +32,30 @@
...
-
-  MultiplexFilterResultRaw::MultiplexFilterResultRaw(double mz, std::vector<double> mz_shifts, std::vector<double> intensities) :
-    mz_(mz), mz_shifts_(mz_shifts), intensities_(intensities)
+  MultiplexSatelliteProfile::MultiplexSatelliteProfile(double rt, double mz, double intensity) :
+    rt_(rt), mz_(mz), intensity_(intensity)
   {
   }
 
-  double MultiplexFilterResultRaw::getMZ() const
+  double MultiplexSatelliteProfile::getRT() const
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexSatelliteProfile.cpp]: No. In centroid satellites I store only indices i.e. `size_t`.
In profile satellites I store coordinates (plus inetnsities) i.e. `double/float`.

File: src/tests/class_tests/openms/source/MultiplexClustering_test.cpp [hunk 21]
```diff
@@ -103,7 +100,7 @@ std::vector<MultiplexFilterResult> filter_results = filtering.filter();
 MultiplexClustering* nullPointer = nullptr;
 MultiplexClustering* ptr;
 
-START_SECTION(MultiplexClustering(const PeakMap& exp_profile, const PeakMap& exp_picked, const std::vector<std::vector<PeakPickerHiRes::PeakBoundary> >& boundaries, double rt_typical, double rt_minimum))
+START_SECTION(MultiplexClustering(const MSExperiment& exp_profile, const MSExperiment& exp_picked, const std::vector<std::vector<PeakPickerHiRes::PeakBoundary> >& boundaries, double rt_typical, double rt_minimum))
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/tests/class_tests/openms/source/MultiplexClustering_test.cpp]: During development, all `Multiplex*` tests were disabled. They are switched on again.

File: src/tests/class_tests/openms/source/MultiplexFilteringCentroided_test.cpp [hunk 22]
```diff
@@ -75,7 +72,7 @@ double mz_tolerance = 40;
 bool mz_tolerance_unit = true;    // ppm (true), Da (false)
 
 // construct list of peak patterns
-MultiplexDeltaMasses shifts1;
+/*MultiplexDeltaMasses shifts1;
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/tests/class_tests/openms/source/MultiplexFilteringCentroided_test.cpp]: Done.

File: src/topp/FeatureFinderMultiplex.cpp [hunk 23]
```diff
@@ -367,9 +375,37 @@ class TOPPFeatureFinderMultiplex :
...
   }
 
+  /**
+   * @brief order of charge states
+   *
+   * 2+ 3+ 4+ 1+ 5+ 6+ ...
+   *
+   * Order charge states by the likelihood of their occurrence, i.e. we search for the most likely charge states first.
+   */
+  static size_t order_charge(int charge)
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/topp/FeatureFinderMultiplex.cpp]: `orderCharge()` now in `FeatureFinderMultiplexAlgorithm.cpp`.
In all use cases so far, that's the order of occurrences.

File: src/topp/FeatureFinderMultiplex.cpp [hunk 24]
```diff
@@ -1044,14 +1573,21 @@ class TOPPFeatureFinderMultiplex :
...
-    consensus_map.setPrimaryMSRunPath(ms_runs);
-
     FeatureMap feature_map;
-    feature_map.setPrimaryMSRunPath(ms_runs);
-
-    generateMaps_(centroided, patterns, filter_results, cluster_results, consensus_map, feature_map);
+    
+    if (centroided)
+    {
+      //consensus_map.setPrimaryMSRunPath(exp_centroid_.getPrimaryMSRunPath());
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/topp/FeatureFinderMultiplex.cpp]: Doesn't compile.

Wanted to check with you how the API has changed. Apart from this path, are other meta data needed for the map?

File: src/topp/FeatureFinderMultiplex.cpp [hunk 25]
```diff
@@ -1061,12 +1597,12 @@ class TOPPFeatureFinderMultiplex :
       writeFeatureMap_(out_features_, feature_map);
     }
 
-    if (out_mzq_ != "")
+    /*if (out_mzq_ != "")
```
[the remark this answers is not in the document]
Comment (lars20070) [reply src/topp/FeatureFinderMultiplex.cpp]: Already removed.

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.h [hunk 26]
```diff
@@ -73,188 +76,133 @@ namespace OpenMS
...
-     * @param pattern    pattern of isotopic peaks to be searched for
-     * @param spectrum_index    index of the spectrum in exp_picked_ and boundaries_
-     * @param mz_shifts_actual_indices    indices of peaks corresponding to the pattern
+     * @param mz    position where the peak is expected
+     * @param mz_tolerance    m/z tolerance within the peak may lie
+     * @param it_rt    pointer to the spectrum
+     * @param intensity_first_peak    intensity to compare to
      *
-     * @return true if all intensities above threshold
+     * @return boolean if there is a significant peak
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.h]: document what a significant peak is

**Comment by timosachsenberg** [comment]
Thanks. I think it is ready for merge as it is a great improvement to the last version (both algorithmic but also internal usability). Please consider the comments for further improvements (e.g. maybe using a different concept for explaining the different outcomes aka white/grey/black)

**Pull request merged**